### DISCUSSION TITLE
importing hyrule

### DISCUSSION THREAD

**Discussion by esakilpi** [hylang/hy#2301] [Q&A] [answered]
I have been trying to import collections.assoc from hyrule, but I always get the same error. Even with (import hyrule).
I guess there is a bug  in macrotools.macroexpansion-all that prevents the import?

This is the error message:
```
hy.errors.HyMacroExpansionError: 

  File ".../lib/python3.8/site-packages/hyrule/macrotools.hy", line 113
    (cond [(in (get form 0) '[unquote unquote-splice])
                      (+quote -1)]
                     [(= (get form 0) 'quasiquote) (+quote)]
                     [True (traverse form)])]
    ^^
expanding macro cond
  TypeError: `cond` needs an even number of arguments
```

**Comment by gabriel-francischini** [comment, marked as answer]
Which versions of Hy and Hyrule are you using?

This seems to work in a virtual env:

```console
(hy-dev) gabriel@linuxstudio:/tmp$ hy
Hy 1.0a4 using CPython(main) 3.10.4 on Linux
=> (require hyrule.collections [assoc])
=> 
=> (do (setv collection {}) (assoc collection "Dog" "Bark") (print collection))
{'Dog': 'Bark'}
=> 
```

Here's the environment:

```console
(hy-dev) gabriel@linuxstudio:/tmp$ pip list
Package          Version
---------------- -------
appdirs          1.4.4
colorama         0.4.4
funcparserlib    1.0.0
funcy            1.17
funcy-chain      0.2.0
hy               1.0a4  <---
hyrule           0.1    <---
multipledispatch 0.6.0
pip              22.1.2
pyrsistent       0.18.1
rply             0.7.8
setuptools       62.3.2
six              1.16.0
toolz            0.11.2
wheel            0.37.1

(hy-dev) gabriel@linuxstudio:/tmp$ conda list python
# packages in environment at /home/superwolf/miniconda3/envs/hy-dev:
#
# Name                    Version                   Build  Channel
python                    3.10.4          h2660328_0_cpython    conda-forge
python_abi                3.10                    2_cp310    conda-forge
```

You can upgrade your versions of hy and hyrule with:
```bash
pip install --upgrade --pre hy hyrule
```

**Reply by Kodiologist** [reply]
Yeah, I think OP is using mismatched versions of Hy and Hyrule.

**Reply by esakilpi** [reply]
I thougth about that, but I think I'm using the latest version of both. Upgrading does not change the versions. Could it be that my python version is too old?

```
pip list | grep hy
hy                           1.0a4+183.gb3920666
hyrule                       0.1

Hy 1.0a4+183.gb3920666 using CPython(default) 3.8.1 on Linux
```

**Reply by Kodiologist** [reply]
It looks like you're using Hy master but Hyrule 0.1. You need Hyrule master.

**Reply by esakilpi** [reply]
I tried to install hyryle master like this but I got version 0.0.0 

`pip install git+https://github.com/hylang/hyrule.git@master`

**Reply by gabriel-francischini** [reply]
IIRC, version 0.0.0 in the master just means it's a yet-to-be-released version. Does the `(require hyrule.collections [assoc])` still throws even with it?

Python 3.8 is almost 3 years old now, but it shouldn't be the issue. You can use [Miniconda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html#regular-installation) (or virtualenv) to replicate the environment that I used, in order to try to rule out issues related to that python version:

```console
source activate base ; conda activate base
conda create --yes --name replicated-hy python=3.10
conda activate replicated-hy
pip install --pre hy hyrule
hy
```

**Reply by esakilpi** [reply]
Yes, it works now. Thank you very much. **import** did'nt  work, it doesn't recognize hy macros, but **require** does. I found the reason in the hy manual.
